// @lrnwebcomponents/hax-body/lib/hax-app-search.js
define(["exports","require","../../../lit-element/lit-element.js","../../simple-colors/simple-colors.js","./hax-store.js","../../../mobx/dist/mobx.esm.js"],(function(_exports,_require,_litElement,_simpleColors,_haxStore,_mobxEsm){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HaxAppSearch=void 0,_require=babelHelpers.interopRequireWildcard(_require);class HaxAppSearch extends _simpleColors.SimpleColors{static get styles(){return[...super.styles,_litElement.css`
        :host {
          display: block;
        }
        hexagon-loader {
          display: none;
          justify-content: center;
          width: 100%;
          z-index: 1000;
          position: absolute;
          --hexagon-color: var(--hax-color-bg-accent, #0085ba);
        }
        hexagon-loader[loading] {
          display: block;
          opacity: 0.8;
        }
        .card-content {
          padding: 16px;
        }
        .card-content p {
          padding: 0;
          margin: 0;
        }
        #itemlist {
          min-height: 172px;
          text-align: center;
          align-items: center;
        }
        hax-app-search-inputs {
          min-height: 80px;
          padding: 0 16px;
        }
        hax-app-pagination {
          min-height: 32px;
          font-size: 12.8px;
          display: none;
          justify-content: flex-end;
          justify-content: center;
        }
        .tos-text {
          font-size: 12px;
        }
        .tos-text ul {
          padding: 4px;
          margin: 0 16px;
        }
        .tos-text a {
          font-size: 12px;
          color: blue;
          text-decoration: underline;
        }
        .tos-text a:hover,
        .tos-text a:focus,
        .tos-text a:active {
          outline: 2px solid blue;
        }
      `]}constructor(){super(),this.auto=!1,this.headers={},this.method="GET",this.loading=!1,this.requestData={},this.media=[],this.tos=[],this.resultMap={},new Promise((res,rej)=>_require.default(["../../simple-fields/lib/simple-fields-field.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../simple-fields/lib/simple-fields-container.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../hexagon-loader/hexagon-loader.js"],res,rej)),new Promise((res,rej)=>_require.default(["./hax-app-search-inputs.js"],res,rej)),new Promise((res,rej)=>_require.default(["./hax-app-search-result.js"],res,rej)),(0,_mobxEsm.autorun)(()=>{this.activeApp=(0,_mobxEsm.toJS)(_haxStore.HAXStore.activeApp)})}render(){return _litElement.html`
      ${this.tos.length>0?_litElement.html`
            <div class="tos-text">Terms of service:</div>
            <ul class="tos-text">
              ${this.tos.map(item=>_litElement.html`
                  <li>
                    <a
                      href="${item.link}"
                      target="_blank"
                      rel="noopener nofollow noreferrer"
                      >${item.title}</a
                    >
                  </li>
                `)}
            </ul>
          `:""}
      <hax-app-search-inputs
        id="searchinput"
        .label="${this.label}"
        .schema="${this.searchSchema}"
        @search-values-changed="${this._searchValuesChanged}"
      ></hax-app-search-inputs>
      <hax-app-pagination
        id="pagerbottom"
        .request-data="${this.requestData}"
        .pagination="${this.pagination}"
      ></hax-app-pagination>
      <hexagon-loader
        size="medium"
        item-count="4"
        ?loading="${this.loading}"
        aria-roledescription="Loading"
      ></hexagon-loader>
      <div id="itemlist">
        ${this.media.map(resultData=>_litElement.html`
            <hax-app-search-result
              image="${resultData.image}"
              title="${resultData.title}"
              details="${resultData.details}"
              .map="${resultData.map}"
              type="${resultData.type}"
            ></hax-app-search-result>
          `)}
      </div>
    `}async loadAppData(){this.loading=!0;let url=this.requestUrl(this.requestEndPoint,this.requestParams);await fetch(url,{headers:this.headers,method:this.method}).then(response=>{if(response.ok)return response.json()}).then(json=>{this._requestDataChanged(json)})}updated(changedProperties){changedProperties.forEach((oldValue,propName)=>{["auto","method","headers","requestEndPoint","requestParams"].includes(propName)&&(clearTimeout(this.__debounce),this.__debounce=setTimeout(()=>{this.requestEndPoint&&this.loadAppData()},100)),"activeApp"==propName&&(this.requestParams={},this.searchSchema={},setTimeout(()=>{this.searchSchema={properties:{}},this._resetAppSearch(this.activeApp)},10))})}requestUrl(url="",params={}){var queryString="";if(null!=_haxStore.HAXStore.connectionRewrites.appendUploadEndPoint&&params.__HAXAPPENDUPLOADENDPOINT__&&(queryString=_haxStore.HAXStore.connectionRewrites.appendUploadEndPoint+"&"),null!=_haxStore.HAXStore.connectionRewrites.appendJwt&&params.__HAXJWT__&&(params[_haxStore.HAXStore.connectionRewrites.appendJwt]=localStorage.getItem(_haxStore.HAXStore.connectionRewrites.appendJwt)),queryString+=this.queryStringData(params)){var bindingChar=url.indexOf("?")>=0?"&":"?";return url+bindingChar+queryString}return url}queryStringData(params){var param,value,queryParts=[];for(param in params)if(value=params[param],"__HAXJWT__"==param||"__HAXAPPENDUPLOADENDPOINT__"==param);else if(Array.isArray(value))for(var i=0;i<value.length;i++)queryParts.push(param+"="+window.encodeURIComponent(value[i]));else null!==value?queryParts.push(param+"="+window.encodeURIComponent(value)):queryParts.push(param);return queryParts.join("&")}static get tag(){return"hax-app-search"}static get properties(){return{activeApp:{type:Object},tos:{type:Array},auto:{type:Boolean},searchSchema:{type:Object},headers:{type:Object},method:{type:String},loading:{type:Boolean},requestData:{type:Object},media:{type:Array},requestEndPoint:{type:String},requestParams:{type:Object},resultMap:{type:Object}}}_searchValuesChanged(e){let requestParams=this.requestParams;for(let property in e.detail)""!=e.detail[property]&&(requestParams[property]=e.detail[property]);this.requestParams={...this.requestParams}}_resetAppSearch(newValue){if(newValue&&newValue.details){let app=newValue;var requestParams={};this.label=app.details.title,app.details.tos&&app.details.tos.length>0?this.tos=[...app.details.tos]:this.tos=[],this.label=app.details.title,this.auto=!1,this.media=[],void 0!==app.connection.data&&(requestParams=app.connection.data),void 0!==app.connection.operations.browse.data&&(requestParams=Object.assign(requestParams,app.connection.operations.browse.data)),this.method=app.connection.operations.browse.method,this.headers={},void 0!==app.connection.headers&&(this.headers=app.connection.headers),this.requestParams={...requestParams};var requestEndPoint=app.connection.protocol+"://"+app.connection.url;"/"!=requestEndPoint.substr(requestEndPoint.length-1)&&(requestEndPoint+="/"),void 0!==app.connection.operations.browse.endPoint&&(requestEndPoint+=app.connection.operations.browse.endPoint),this.requestEndPoint=requestEndPoint;var searchSchema={properties:{}};void 0!==app.connection.operations.browse.search&&(searchSchema.properties=app.connection.operations.browse.search,this.searchSchema={...searchSchema}),this.resultMap=app.connection.operations.browse.resultMap,this.pagination={},void 0!==app.connection.operations.browse.pagination&&(this.pagination=app.connection.operations.browse.pagination),void 0!==app.connection.auto?this.auto=app.connection.auto:this.auto=!0}}_requestDataChanged(newValue){if(this.resultMap&&typeof newValue!={}){let media=[],map=this.resultMap,data=[];if(this.resultMap.items?void 0!==this._resolveObjectPath(map.items,newValue)?data=this._resolveObjectPath(map.items,newValue):null!=newValue&&(data=newValue):data=newValue,null!=data){for(var i=0;i<data.length;i++){if(media[i]={title:this._resolveObjectPath(map.preview.title,data[i]),details:this._resolveObjectPath(map.preview.details,data[i]),type:map.defaultGizmoType,map:{}},void 0!==media[i].details&&null!=media[i].details&&(media[i].details=media[i].details.replace(/(<([^>]+)>)/gi,"")),map.preview.id.constructor===Object){let tmp=this._resolveObjectPath(map.preview.id.property,data[i]);"split"===map.preview.id.op&&(tmp=tmp.split(map.preview.id.delimiter),media[i].id=tmp[map.preview.id.position])}else media[i].id=this._resolveObjectPath(map.preview.id,data[i]);for(var prop in void 0!==map.preview.image?media[i].image=this._resolveObjectPath(map.preview.image,data[i]):void 0!==map.image?media[i].image=map.image:media[i].image="",map.gizmo)if("_url_source"===prop){let _id="";_id=void 0!==media[i].map.__id?media[i].map.__id:this._resolveObjectPath(map.gizmo.id,data[i]),media[i].map.source=map.gizmo._url_source.replace("<%= id %>",_id),media[i].map.url=media[i].map.source}else if(map.gizmo[prop].constructor===Object){let tmp=this._resolveObjectPath(map.gizmo[prop].property,data[i]);"split"===map.gizmo[prop].op&&(tmp=tmp.split(map.gizmo[prop].delimiter),media[i].map[prop]=tmp[map.gizmo[prop].position],"id"===prop&&(media[i].map.__id=media[i].map[prop]))}else media[i].map[prop]=this._resolveObjectPath(map.gizmo[prop],data[i]);void 0===media[i].map.url&&void 0!==media[i].map.source&&(media[i].map.url=media[i].map.source),void 0!==map.gizmo.type?media[i].type=this._resolveObjectPath(map.gizmo.type,data[i]):void 0!==map.gizmo.mimetype?media[i].type=_haxStore.HAXStore.mimeTypeToGizmoType(this._resolveObjectPath(map.gizmo.mimetype,data[i])):"*"!=_haxStore.HAXStore.guessGizmoType(map.gizmo)&&(media[i].type=_haxStore.HAXStore.guessGizmoType(map.gizmo))}this.media=[...media]}}this.loading=!1}_resolveObjectPath(path,obj){return path.split(".").reduce((function(prev,curr){return prev?prev[curr]:null}),obj||self)}}_exports.HaxAppSearch=HaxAppSearch,window.customElements.define(HaxAppSearch.tag,HaxAppSearch)}));
