// @lrnwebcomponents/hax-body/lib/hax-map.js
define(["exports","../../../lit-element/lit-element.js","../../simple-icon/lib/simple-icon-lite.js","../../simple-icon/lib/simple-icon-button.js","../../hax-iconset/lib/simple-hax-iconset.js","./hax-store.js","../../utils/utils.js"],(function(_exports,_litElement,_simpleIconLite,_simpleIconButton,_simpleHaxIconset,_haxStore,_utils){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HaxMap=void 0;class HaxMap extends _litElement.LitElement{static get styles(){return[_litElement.css`
        :host {
          display: block;
        }
        .title {
          position: relative;
          padding: 16px;
          outline: 0;
          font-weight: 600;
          text-align: left;
          margin: 0;
          background-color: var(--hax-color-menu-heading-bg, #eeeeee);
          color: var(--hax-color-menu-heading-color, black);
          font-size: 18px;
          line-height: 18px;
          font-family: "Noto Serif", serif;
        }
        .container {
          text-align: left;
        }
        table {
          font-size: 13px;
        }
        table caption {
          font-weight: bold;
        }
        table tr th {
          padding: 2px;
        }
        table td {
          font-size: 21px;
          font-weight: bold;
          text-align: center;
        }
        h4 {
          font-size: 14px;
          margin: 2px 0;
        }
        ul {
          list-style: none;
          padding: 0;
          margin: 0;
        }
        ul li {
          margin: 4px;
          padding: 0;
        }
        ul simple-icon-lite {
          padding: 0 8px;
        }
        a {
          font-size: 24px;
          line-height: 24px;
          text-decoration: none;
          color: black;
          padding: 4px;
          display: block;
        }
        a:focus,
        a:hover,
        a:active {
          cursor: pointer;
          font-weight: bold;
          outline: 2px solid black;
        }
        simple-icon-button {
          float: right;
        }
      `]}constructor(){super(),this.elementList=[],this.title="Content map"}updateHAXMap(){let list=_haxStore.HAXStore.htmlToHaxElements(_haxStore.HAXStore.activeHaxBody.haxToContent());this.calcStats(list);let elements=[];for(var i=0;i<list.length;i++){let def=_haxStore.HAXStore.haxSchemaFromTag(list[i].tag);def.gizmo?elements.push({icon:def.gizmo.icon,name:def.gizmo.title}):list[i].tag&&list[i].tag.includes("-")?elements.push({icon:"hax:templates",name:"Widget"}):elements.push({icon:"hax:paragraph",name:"HTML block"})}this.elementList=[...elements]}calcStats(elements){if(elements&&_haxStore.HAXStore.activeHaxBody.innerText){let counts={c:_haxStore.HAXStore.activeHaxBody.innerText.length,w:parseInt(_haxStore.HAXStore.activeHaxBody.innerText.split(/\s+/g).length-1),h:0,p:0,e:0};for(var i in elements.forEach(el=>{switch(el.tag){case"blockquote":case"div":case"span":case"p":counts.p++;break;case"h1":case"h2":case"h3":case"h4":case"h5":case"h6":case"relative-heading":counts.h++;break;default:counts.e++}}),counts)this[`${i}Count`]=counts[i]}}closeBtn(e){this.dispatchEvent(new CustomEvent("hax-tray-button-click",{bubbles:!0,cancelable:!0,composed:!0,detail:{eventName:"open-map",index:0,value:!0}}))}render(){return _litElement.html`
      <h3 class="title">
        <simple-icon-lite icon="hax:map"></simple-icon-lite>
        ${this.title}
        <simple-icon-button
          icon="close"
          @click="${this.closeBtn}"
        ></simple-icon-button>
      </h3>
      <div class="container">
        <table>
          <caption>
            Content statistics
          </caption>
          <tr>
            <th>Words</th>
            <th>Headings</th>
            <th>Paragraphs</th>
            <th>Widgets</th>
            <th>Characters</th>
          </tr>
          <tr>
            <td>${this.wCount}</td>
            <td>${this.hCount}</td>
            <td>${this.pCount}</td>
            <td>${this.eCount}</td>
            <td>${this.cCount}</td>
          </tr>
        </table>
        <h4>List view</h4>
        <ul>
          ${this.elementList.map((element,index)=>_litElement.html`
              <li>
                <a @click="${this.scrollInMap}" data-index="${index}"
                  ><simple-icon-lite
                    data-index="${index}"
                    icon="${element.icon}"
                  ></simple-icon-lite
                  >${element.name}</a
                >
              </li>
            `)}
        </ul>
      </div>
    `}scrollInMap(e){var target=(0,_utils.normalizeEventPath)(e)[0];if(target.getAttribute("data-index")){let activeChild=_haxStore.HAXStore.activeHaxBody.children[parseInt(target.getAttribute("data-index"))];activeChild.classList.add("blinkfocus"),"function"==typeof activeChild.scrollIntoViewIfNeeded?activeChild.scrollIntoViewIfNeeded(!0):activeChild.scrollIntoView({behavior:"smooth",inline:"center"}),setTimeout(()=>{activeChild.classList.remove("blinkfocus")},500)}}static get tag(){return"hax-map"}static get properties(){return{title:{type:String},opened:{type:Boolean},elementList:{type:Array},cCount:{type:String},wCount:{type:String},hCount:{type:String},pCount:{type:String},eCount:{type:String}}}}_exports.HaxMap=HaxMap,window.customElements.define(HaxMap.tag,HaxMap)}));
