// @lrnwebcomponents/simple-modal/simple-modal.js
/**
 * Copyright 2019 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{LitElement,html,css}from"../../lit-element/lit-element.js";import"../simple-icon/lib/simple-icons.js";import"../simple-icon/lib/simple-icon-lite.js";import"../simple-icon/lib/simple-icon-button-lite.js";import"../../web-dialog/index.js";class SimpleModal extends LitElement{static get styles(){return[css`
        :host {
          display: block;
        }

        :host([hidden]) {
          display: none;
        }

        :host web-dialog ::slotted(*) {
          font-size: 14px;
          width: 100%;
        }

        #titlebar {
          margin-top: 0;
          padding: var(--simple-modal-titlebar-padding, 0px 16px);
          display: flex;
          align-items: center;
          justify-content: space-between;
          color: var(--simple-modal-titlebar-color, #444);
          background-color: var(--simple-modal-titlebar-background, #ddd);
          border-radius: 0;
          height: var(--simple-modal-titlebar-height, unset);
          line-height: var(--simple-modal-titlebar-line-height, unset);
        }

        #headerbar {
          margin: 0;
          padding: var(--simple-modal-header-padding, 0px 16px);
          color: var(--simple-modal-header-color, #222);
          background-color: var(--simple-modal-header-background, #ccc);
        }

        h2 {
          margin-right: 8px;
          padding: 0;
          margin: 0;
          flex: auto;
          font-size: 18px;
          line-height: 18px;
        }

        #close {
          top: 0;
          border: var(--simple-modal-titlebar-button-border, none);
          padding: var(--simple-modal-titlebar-button-padding, 10px 0);
          min-width: unset;
          text-transform: none;
          color: var(--simple-modal-titlebar-color, #444);
          background-color: transparent;
        }

        #close:focus {
          opacity: 0.7;
          outline: var(--simple-modal-titlebar-button-outline, 2px dotted grey);
          outline-offset: var(
            --simple-modal-titlebar-button-outline-offset,
            2px
          );
        }

        #close simple-icon-lite {
          --simple-icon-height: var(--simple-modal-titlebar-icon-height, 16px);
          --simple-icon-width: var(--simple-modal-titlebar-icon-width, 16px);
          color: var(--simple-modal-titlebar-color, #444);
        }

        #simple-modal-content {
          flex-grow: 1;
          padding: var(--simple-modal-content-padding, 8px 16px);
          margin: 0;
          color: var(--simple-modal-content-container-color, #222);
          background-color: var(
            --simple-modal-content-container-background,
            #fff
          );
        }

        .buttons {
          padding: 0;
          padding: var(--simple-modal-buttons-padding, 0);
          margin: 0;
          color: var(--simple-modal-buttons-color, unset);
          background-color: var(--simple-modal-buttons-background, unset);
        }

        .buttons ::slotted(*) {
          padding: 0;
          margin: 0;
          color: var(--simple-modal-button-color, --simple-modal-buttons-color);
          background-color: var(
            --simple-modal-button-background,
            --simple-modal-buttons-background
          );
        }
        web-dialog {
          --dialog-border-radius: var(--simple-modal-border-radius, 2px);
          z-index: var(--simple-modal-z-index, 1) !important;
        }
        web-dialog::part(dialog) {
          border: 1px solid var(--simple-modal-border-color, #222);
          padding: 0;
        }
        web-dialog.style-scope.simple-modal {
          display: none !important;
        }
        web-dialog[open].style-scope.simple-modal {
          display: flex !important;
          position: fixed !important;
          top: 25vh;
          left: 25vw;
        }
      `]}render(){return html`<web-dialog
      id="dialog"
      center
      role="dialog"
      aria-describedby="simple-modal-content"
      aria-label="${this._getAriaLabel(this.title)}"
      aria-labelledby="${this._getAriaLabelledby(this.title)}"
      aria-modal="true"
      ?open="${this.opened}"
      @open="${this.open}"
      @close="${this.close}"
    >
      <div id="titlebar">
        <h2 id="simple-modal-title" ?hidden="${!this.title}">${this.title}</h2>
        <div></div>
        <simple-icon-button-lite
          id="close"
          dark
          icon="${this.closeIcon}"
          @click="${this.close}"
          label="${this.closeLabel}"
        >
        </simple-icon-button-lite>
      </div>
      <div id="headerbar"><slot name="header"></slot></div>
      <div id="simple-modal-content">
        <slot name="content"></slot>
      </div>
      <div class="buttons">
        <slot name="buttons"></slot>
      </div>
    </web-dialog>`}static get properties(){return{...super.properties,title:{type:String},opened:{type:Boolean,reflect:!0},closeLabel:{attribute:"close-label",type:String},closeIcon:{type:String,attribute:"close-icon"},invokedBy:{type:Object},modal:{type:Boolean}}}static get tag(){return"simple-modal"}constructor(){super(),this.title="",this.opened=!1,this.closeLabel="Close",this.closeIcon="close",this.modal=!1}updated(changedProperties){changedProperties.forEach((oldValue,propName)=>{"opened"==propName&&this._openedChanged(this[propName])})}connectedCallback(){super.connectedCallback(),setTimeout(()=>{window.addEventListener("simple-modal-hide",this.close.bind(this)),window.addEventListener("simple-modal-show",this.showEvent.bind(this))},0)}disconnectedCallback(){window.removeEventListener("simple-modal-hide",this.close.bind(this)),window.removeEventListener("simple-modal-show",this.showEvent.bind(this)),super.disconnectedCallback()}showEvent(e){window.dispatchEvent(new CustomEvent("simple-toast-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1})),this.opened?(this.innerHTML="",setTimeout(()=>{this.show(e.detail.title,e.detail.elements,e.detail.invokedBy,e.detail.id,e.detail.modalClass,e.detail.styles,e.detail.clone,e.detail.modal)},0)):this.show(e.detail.title,e.detail.elements,e.detail.invokedBy,e.detail.id,e.detail.modalClass,e.detail.styles,e.detail.clone,e.detail.modal)}show(title,elements,invokedBy,id=null,modalClass=null,styles=null,clone=!1,modal=!1){let element;this.invokedBy=invokedBy,this.modal=modal,this.title=title;let slots=["header","content","buttons"];for(var i in id?this.setAttribute("id",id):this.removeAttribute("id"),this.setAttribute("style",""),styles&&["--simple-modal-width","--simple-modal-z-index","--simple-modal-height","--simple-modal-min-width","--simple-modal-min-height","--simple-modal-max-width","--simple-modal-max-height","--simple-modal-titlebar-color","--simple-modal-titlebar-height","--simple-modal-titlebar-line-height","--simple-modal-titlebar-background","--simple-modal-titlebar-padding","--simple-modal-header-color","--simple-modal-header-background","--simple-modal-header-padding","--simple-modal-content-container-color","--simple-modal-content-container-background","--simple-modal-content-padding","--simple-modal-buttons-color","--simple-modal-buttons-background","--simple-modal-buttons-padding","--simple-modal-button-color","--simple-modal-button-background"].forEach(prop=>{this.style.setProperty(prop,styles[prop]||"inherit")}),modalClass?this.setAttribute("class",modalClass):this.removeAttribute("class"),slots)elements[slots[i]]&&(element=clone?elements[slots[i]].cloneNode(!0):elements[slots[i]],element.setAttribute("slot",slots[i]),this.appendChild(element));this.opened=!0}close(){this.opened=!1,window.ShadyCSS&&!window.ShadyCSS.nativeShadow&&(this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.position="relative")}open(){this.opened=!0,window.ShadyCSS&&!window.ShadyCSS.nativeShadow&&(this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.position="fixed",this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.top=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.bottom=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.left=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.right=0)}_openedChanged(newValue){if(void 0===newValue||newValue){if(newValue){this.querySelectorAll("[dialog-dismiss],[dialog-confirm]").forEach(el=>{el.addEventListener("click",e=>{this.close()})});const evt=new CustomEvent("simple-modal-opened",{bubbles:!0,cancelable:!0,detail:{opened:!0,invokedBy:this.invokedBy}});this.dispatchEvent(evt)}}else{for(this.title="";null!==this.firstChild;)this.removeChild(this.firstChild);this.invokedBy&&setTimeout(()=>{this.invokedBy.focus()},500);const evt=new CustomEvent("simple-modal-closed",{bubbles:!0,cancelable:!0,detail:{opened:!1,invokedBy:this.invokedBy}});this.dispatchEvent(evt)}}_getAriaLabelledby(title){return title?"simple-modal-title":null}_getAriaLabel(title){return title?null:"Modal Dialog"}}window.customElements.define(SimpleModal.tag,SimpleModal);export{SimpleModal};window.SimpleModal=window.SimpleModal||{},window.SimpleModal.requestAvailability=()=>(window.SimpleModal.instance||(window.SimpleModal.instance=document.createElement("simple-modal"),document.body.appendChild(window.SimpleModal.instance)),window.SimpleModal.instance);export const SimpleModalStore=window.SimpleModal.requestAvailability();
