// @lrnwebcomponents/accent-card/lib/accent-card-clickable.js
define(["exports","../../../lit-element/lit-element.js","../../simple-colors/simple-colors.js","../accent-card.js"],(function(_exports,_litElement,_simpleColors,_accentCard){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.AccentCardClickable=void 0;
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class AccentCardClickable extends _litElement.LitElement{static get tag(){return"accent-card-clickable"}static get styles(){return[_litElement.css`
        :host,
        accent-card {
          display: flex;
          align-items: stretch;
          justify-content: stretch;
          flex-direction: column;
        }
        accent-card {
          position: relative;
          flex: 1 1 auto;
        }
        accent-card:focus-within,
        accent-card:hover {
          outline: 1px solid
            var(
              --accent-card-border-color,
              --simple-colors-default-theme-accent-6,
              #ddd
            );
        }
        ::slotted(*:not([data-clickable]:first-of-type)) {
          z-index: 2;
          position: relative;
        }
        ::slotted([data-clickable]:first-of-type) {
          z-index: 1;
        }
        ::slotted([data-clickable]:first-of-type):after {
          content: " ";
          position: absolute;
          left: 0;
          top: 0;
          bottom: 0;
          right: 0;
        }
      `]}render(){return _litElement.html`
      <accent-card
        accent-color="${this.accentColor}"
        ?accent-background="${this.accentBackground}"
        ?accent-heading="${this.accentHeading}"
        ?flat="${this.flat}"
        ?horizontal="${this.horizontal}"
        ?no-border="${this.noBorder}"
        image-align="${this.imageAlign}"
        image-src="${this.imageSrc}"
        image-valign="${this.imageValign}"
      >
        <slot name="heading" slot="heading"></slot>
        <slot name="corner" slot="corner"></slot>
        <slot name="subheading" slot="subheading"></slot>
        <slot name="content" slot="content"></slot>
        <slot name="footer" slot="footer"></slot>
      </accent-card>
    `}static get haxProperties(){_accentCard.AccentCard.haxProperties}constructor(){super(),this.accentBackground=!1,this.accentHeading=!1,this.flat=!1,this.horizontal=!1,this.noBorder=!1}static get properties(){return{..._simpleColors.SimpleColors.properties,..._accentCard.AccentCard.properties}}}_exports.AccentCardClickable=AccentCardClickable,window.customElements.define(AccentCardClickable.tag,AccentCardClickable)}));
