// @lrnwebcomponents/simple-colors/lib/demo/simple-colors-swatches.js
define(["exports","require","../../../../lit-element/lit-element.js","../../simple-colors.js"],(function(_exports,_require,_litElement,_simpleColors){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.simpleColorsSwatches=void 0,_require=babelHelpers.interopRequireWildcard(_require);
/**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class simpleColorsSwatches extends _simpleColors.SimpleColors{static get styles(){return[...super.styles,_litElement.css`
        :host {
          display: block;
          margin: 15px 0;
        }
        :host([hidden]) {
          display: none;
        }
        :host simple-modal-template {
          --simple-modal-titlebar-color: var(--simple-colors-swatch-contrast);
          --simple-modal-titlebar-background: var(--simple-colors-swatch-color);
        }
        :host .row {
          display: flex;
          align-items: stretch;
        }
        :host button {
          width: calc(8% - 6px);
          height: 80px;
          margin: 3px;
          border: 1px solid black;
          border-radius: 3px;
          padding: 0;
          display: flex;
          flex-direction: column;
          justify-content: flex-end;
          align-items: stretch;
        }
        :host button div {
          margin: 0;
          padding: 2px;
          font-size: 10px;
          background: black;
          color: white;
          width: calc(100% - 4px);
        }
        :host button:focus,
        :host button:hover {
          opacity: 0.7;
          cursor: pointer;
        }
      `]}render(){return _litElement.html`
      <div>
        <p>Click on the swatches above to learn more.</p>
        <div class="rows">
          ${Object.keys(this.colors).map(color=>_litElement.html`
              <div class="row">
                ${this.colors.grey.map((shade,index)=>_litElement.html`
                    <button
                      id="${this._getId(color,index)}"
                      controls="swatchinfo"
                      style="background-color: ${this.colors[color][index]};"
                      title="${this._getTitle(color,index)}"
                      @click="${e=>this._handleClick(color,index)}"
                    >
                      <div>${this._getTitle(color,index)}</div>
                    </button>
                  `)}
              </div>
            `)}
        </div>
        <simple-modal-template id="modal" title="${this.swatchName}">
          <p slot="content">
            <simple-colors-swatch-info
              swatch-id="${this.swatchId}"
              swatch-name="${this.swatchName}"
            >
            </simple-colors-swatch-info>
          </p>
        </simple-modal-template>
      </div>
    `}static get properties(){return{swatchId:{attribute:"swatch-id",type:String},swatchName:{attribute:"swatch-name",type:String}}}static get tag(){return"simple-colors-swatches"}updated(changedProperties){changedProperties.forEach((oldValue,propName)=>{"swatchId"===propName&&this.shadowRoot.querySelector("#modal")&&this.shadowRoot.querySelector("#modal").openModal&&this.shadowRoot.querySelector("#modal").openModal(this.shadowRoot.querySelector("#"+this.swatchId))})}constructor(){super(),this.swatchId="grey_0",this.swatchName="grey-1",new Promise((res,rej)=>_require.default(["./simple-colors-swatch-info.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../simple-modal/lib/simple-modal-template.js"],res,rej))}_getTitle(color,index){return`${color}-${this._getShade(index)}`}_getId(color,index){return`${color}_${index}`}_getShade(index){return parseInt(index)+1}_handleClick(color,index){this.swatchId=this._getId(color,index),this.swatchName=this._getTitle(color,index)}}_exports.simpleColorsSwatches=simpleColorsSwatches,window.customElements.define(simpleColorsSwatches.tag,simpleColorsSwatches)}));
